docs/proposals/dedicated-nodes.md

```diff
@@ -0,0 +1,294 @@
+1. (This is the one step is that is specific to dedicated nodes)
+Implement an admission controller that adds tolerations to pods that are supposed
+to be allowed to use dedicated nodes (for example, based on pod's namespace).
+
+In the future one can imagine a generic policy configuration that configures
+an admission controller to apply the appropriate tolerations to the desired class of pods and
+configures the NodeController or another component to apply the appropriate
+taints to the desired set of nodes. It could be used not just for policies about dedicated nodes,
+but also other uses of taints and tolerations, e.g. nodes that are restricted
+due to their hardware configuration.
+
+The `kubectl` command to add and remove taints on nodes will be modeled after `kubectl label`.
+Examples usages:
+
+```sh
+// Update node 'foo' with a taint with key 'dedicated' and value 'special-user' and effect 'NoScheduleNoAdmitNoExecute'.
+// If a taint with that key already exists, its value and effect are replaced as specified.
+$ kubectl taint nodes foo dedicated=special-user=NoScheduleNoAdmitNoExecute
+
+// Remove from node 'foo' the taint with key 'dedicated' if one exists.
+$ kubectl taint nodes foo dedicated-
+```
+
+## Example: implementing a dedicated nodes policy
+
+Let's say that the cluster administrator wants to make nodes `foo`, `bar`, and `baz` available
+only to pods in a particular namespace `banana`. First the administrator does
+
+```sh
+$ kubectl taint nodes foo dedicated=banana=NoScheduleNoAdmitNoExecute
+$ kubectl taint nodes bar dedicated=banana=NoScheduleNoAdmitNoExecute
+$ kubectl taint nodes baz dedicated=banana=NoScheduleNoAdmitNoExecute
+
+```
+
+(assuming they want to evict pods that are already running on those nodes if those
+pods don't already tolerate the new taint)
+
+Then they ensure that the `PodSpec` for all pods created in namespace `banana` specify
+a toleration with `key=dedicated`, `value=banana`, and `policy=NoScheduleNoAdmitNoExecute`.
+
+In the future, it would be nice to be able to specify the nodes via a `NodeSelector` rather than having
+to enumerate them by name.
+
+## Future work: security and quota
+
+At present, the Kubernetes security model allows any user to add and remove any taints and tolerations.
+Obviously this makes it impossible to securely enforce
+rules like dedicated nodes. We need some mechanism that prevents regular users from mutating the `Taints`
+field of `NodeSpec` (probably we want to prevent them from mutating any fields of `NodeSpec`)
+and from mutating the `Tolerations` field of their pods. #17549 is relevant.
+
+Another security vulnterability arises if nodes are added to the cluster before receiving
+their taint. Thus we need to ensure that a new node does not become "Ready" until it has been
+configured with its taints. This is why we suggested earlier that NodeController should be
+responsible for adding taints to nodes.
```

Comment:
How would NodeController decide what taints to apply?

Much like I've proposed Kubelet exporting its own labels, I could imagine it exporting some taints as well.